[Cheng Geng] Duke Increments

ChengGeng97 opened the pull request:
Duke Increments for Week 3.

ChengGeng97 committed:
Added functionality for Level-1.

ChengGeng97 committed:
Added functionality for Level-2. Added some comments to the code.

ChengGeng97 committed:
Added functionality for Level-3

ChengGeng97 committed:
Addded functionality for Level-4

ChengGeng97 committed:
Added functionality for Level-5.

ChengGeng97 committed:
Added functionality for Level-6

ChengGeng97 committed:
Added functionality for Level-7.
- Created a DukeSaveLoad class to handle the saving and loading of data
- Created "toSaveFileString" methods for Task class to help with saving and loading of data
- Edited Duke to save data whenever its internal list changes

ChengGeng97 committed:
Added functionality for Level-8
- Created DukeDateTime class to handle all date and time computation
- Edited Duke's input parsing to work with specific formats for date and time
- Edited EventTask and DeadlineTask to use DukeDateTime to encode timings
- Fixed a bug where TaskList would use the user input index rather than the actual index to complete tasks

ChengGeng97 committed:
Merge branch 'Level-8'

ChengGeng97 committed:
Refactored code to be cleaner + better fitting to OOP standards.
- Decision-making moved out of Duke into UserInputProcessor
- Created TextToTaskTranslator to parse Text inputs into Tasks (Text -> Task translation no longer part of Decision-Making)
- Duke's text lines moved out of Duke into DukeUi
- Deleted all custom Exceptions other than DukeException
- All DukeExceptions, rather than having a hard-coded error message, will call upon the appropriate error message in DukeUi during instantiation.
- Changed DukeSaveLoad (and all saving and loading operations to use "Serializable" rather than String-parsing)
- Changed Task from an Interface to an Abstract Class, changed each Task accordingly
- Changed implementation of DukeDateTime and created
- Created DukeDuration class
- EventTask uses DukeDuration to keep track of its timing
- Added "Nuke" functionality to Duke to allow it to destroy all Tasks in the list

ChengGeng97 committed:
Implemented Increment A-JUnit
- Ported project over into Maven in order to use JUnit
- Created tests for DukeReply and ToDoTask classes

ChengGeng97 committed:
Added functionality for Level-9
- Added ability for UserInputProcessor to recognise and parse 'Find' command
  - Added 'processFindCase' method
  - Added 'Find' to userInputType enum
- Added FEEDBACK_FIND and FEEDBACK_FIND_NOTHING Strings to DukeUi.
- Added 'descriptionContainsTerm' method to Task
- Added 'getMatchingTasksAsString' method to TaskList

ChengGeng97 committed:
Refactored code to be closer to the Java Coding Standard:
https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html

ChengGeng97 committed:
Revert "Refactored code to be closer to the Java Coding Standard:"

This reverts commit 2482a40fab10d81d6589cd6142d4caaa9d1042bb.

ChengGeng97 committed:
Refactored code to abide more closely to module Coding Standard:
https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html

ChengGeng97 committed:
Added javadoc comments for all classes in the project.

ChengGeng97 committed:
Merge tag 'A-JavaDoc'

ChengGeng97 committed:
Merge tag 'A-CodingStandard'

# Conflicts:
#	target/classes/TaskList.class
#	target/classes/TextToTaskTranslator.class
#	target/classes/UserInputProcessor.class

ChengGeng97 committed:
Merge tag 'Level-9'

# Conflicts Resolved:
#	src/main/java/DukeUi.java
#	src/main/java/TaskList.java
#	target/classes/DukeUi.class
#	target/classes/Task.class
#	target/classes/TaskList.class
#	target/classes/UserInputProcessor$userInputType.class
#	target/classes/UserInputProcessor.class

ChengGeng97 committed:
Housekeeping
- Implemented serialVersionUID for Serializable classes
- Added comments explaining use of private methods
- Create javadoc comments for Level-9 methods (related to Find functionality)
- Abstracted out checkIndexExists method to check for exceptions in TaskList
- Replaced using equals("") with using isEmpty() to check for empty Strings

ChengGeng97 committed:
Added a drawing of Duke.

ChengGeng97 committed:
Revert "Added a drawing of Duke."

This reverts commit 7e54b7983f3f73a5cd6f6810706f8a4a16e7bef0.

ChengGeng97 committed:
Added a drawing of Duke.
Robot waifus are the future.

ChengGeng97 committed:
port project over from Maven to Gradle

JunHongT commented:
Like your coding in general. The naming of the variables is descriptive, allowing me to understand the general flow without needing to search for each method to discover more about it. Process had also been segregated to several smaller and concise methods, allowing for minimal repeat of coding. 
However, though i was able to understand the idea of your methods, it might be a great idea to include javadoc for the methods, instead of a 1-liner comment? It might be a huge hassle but i will believe that it will make it clearer for the someone who had no idea about the project to have a better understanding.

JunHongT commented on src/main/java/ToDoTask.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    //Returns a copy of this task but with its completion status marked as undone
+    public Task getTaskMarkedUndone() {
```
May i ask about the situation of when we are required to use this method? Is it an addition function for the user or...
?

ChengGeng97 replied:
It's meant to be an additional feature, for cases where you accidentally marked something as done when it's not actually done. It hasn't been fully implemented though, which is a blunder on my part for having pushed it.

ChengGeng97 commented:
> Like your coding in general. The naming of the variables is descriptive, allowing me to understand the general flow without needing to search for each method to discover more about it. Process had also been segregated to several smaller and concise methods, allowing for minimal repeat of coding.
> However, though i was able to understand the idea of your methods, it might be a great idea to include javadoc for the methods, instead of a 1-liner comment? It might be a huge hassle but i will believe that it will make it clearer for the someone who had no idea about the project to have a better understanding.

Thanks for taking the time to read through my code! Javadoc comments were kept to only the public methods for this build since only those would be generated for the Javadoc. I understand the sentiment though, and will get around to adding them for all private methods as well if time permits.

ChengGeng97 committed:
Implement JavaFX
- Followed the tutorials.
- The cute animals are placeholders. Don't get used to them.

tanamaroby commented on src/main/java/DukeTextFormatter.java:
```diff
@@ -0,0 +1,37 @@
+    private static final String BAR = "____________________________________________________________\n";
+    private static final String FOUR_SPACE = "    ";
+    private static final String FIVE_SPACE = "     ";
+
```
Very epic! Nice work!

tanamaroby replied:
Cool use of private static final attributes!

tanamaroby commented on src/main/java/DukeUi.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * A class that stores all of Duke's mesages as Strings. These are split into Greetings, Feedback and Error Messages.
+ */
+public class DukeUi {
```
This is very great! Nice one ChengGeng97!

tanamaroby reviewed (commented):
Overall your code has a little something for everyone. Great use of Object Oriented Principles which allows reviewers like me to read your code with much ease, that and the descriptive Javadoc really helped. The methods and variables are named appropriately which is descriptive enough that the code is readable on its own even without the use of Javadoc. Great use of DukeUi class in which by using all the final static attributes, you can maintain constant reply from Duke and helps with the consistency of the program. 
One criticism that I would make would be that some of the classes are too spread out and too small that most of them just feels unnecessary in my opinion. For example like the DukeTextFormatter class which I understand is used to maintain constant format of the Duke messages but I feel like this could be combined with the UI class since the logic would be similar. However, this is just my opinion and would not really affect the overall efficiency of the program. 

Conclusion : This code really makes you feel like a software engineer.

ChengGeng97 committed:
Merge branch 'Week-4'

ChengGeng97 committed:
Added Assertions
- Added two assertions in descriptionContainsTerm method for abstract class Task.

ChengGeng97 committed:
Merge pull request #4 from ChengGeng97/A-Assertions

Added Assertions

ChengGeng97 committed:
Made changes to ensure Code Quality
- Deleted unused, commented out code
- Deprecated the DukeTextFormatter, since the project has moved from console to GUI
- Abstracted out checkCommandIncludesIndex method in UserInputProcessor
- Abstracted out bothStartAndEndDateExist boolean in translateTaskEvent method for better readability.

ChengGeng97 committed:
Merge remote-tracking branch 'origin/master'

ChengGeng97 committed:
Added functionality for Notes
- Added commands for adding, deleting and display notes
- Duke now holds a reference to a NoteList
- Added relevant lines to DukeUi
- DukeSaveLoad saves 2 separate files - a NoteList and a TaskList

ChengGeng97 committed:
Add support for ShadowJar so that standalone jar file will include JavaFX files
- Edited build.gradle

ChengGeng97 committed:
Finalise Features for Individual Project
- Revamped the UI to look better and include different portrait photos for Duke
- Added classes GlobalDukeImageChoiceBuffer and public enum DukeImageChoice to improve presentation of the app
- Implemented checkstyle
- Added Javadoc comments for all private methods in addition to all public methods

ChengGeng97 committed:
Delete docs file

ChengGeng97 committed:
Add representative screenshot to the repo

ChengGeng97 committed:
Added README to docs folder

ChengGeng97 committed:
Added User Guide
- Wrote up the User Guide, added example screenshots
- Changed TaskList method getMatchingTasksAsString

ChengGeng97 committed:
Change README file to get pictures to show up

ChengGeng97 committed:
Change README file to get pictures to show up (attempt 2)

ChengGeng97 committed:
Edit README to look nicer

ChengGeng97 committed:
Some final fixes
- Pushed an empty save file for NoteList
- Changed the Readme User Guide
- Removed the default Readme to try to get the User Guide to show up on the root page instead

ChengGeng97 committed:
Make some final edits
- Updated for cross-platform capability with Linux
- Bugfix for inputting leap days (29th of February)
- Added newlines to some files that did not have them
